### PR TITLE
handle avro

### PR DESCRIPTION
**Pull request by ememmr**

### CODE REVIEW

File: client/src/main/scala/sbtcompendium/client/CompendiumClient.scala [hunk 1]
```diff
@@ -16,12 +16,16 @@
...
+import avrohugger.Generator
+import avrohugger.format.Standard
 import cats.effect.Sync
 import cats.free.Free
 import cats.implicits._
 import hammock._
 import hammock.circe.implicits._
 import higherkindness.compendium.models._
+//import higherkindness.droste.data.Mu
+//import higherkindness.skeuomorph.protobuf.ProtobufF
```
Comment (juanpedromoreno) [inline client/src/main/scala/sbtcompendium/client/CompendiumClient.scala]: We could remove these commented lines

File: client/src/main/scala/sbtcompendium/client/CompendiumClient.scala [hunk 2]
```diff
@@ -102,7 +106,17 @@ object CompendiumClient {
         } yield out
       }
 
-      override def generateClient(target: IdlName, identifier: String): F[String] = F.pure("")
+      override def generateClient(target: IdlName, identifier: String): F[List[String]] =
+        target match {
+          case IdlName.Avro =>
+            retrieveProtocol(identifier, None)
```
Comment (juanpedromoreno) [inline client/src/main/scala/sbtcompendium/client/CompendiumClient.scala]: It could be in a separate PR, but the protocols to retrieve should be specified here. We'll probably want to tweak the `compendiumProtocolIdentifiers` sbt setting, replacing:

* `compendiumProtocolIdentifiers: SettingKey[Seq[String]]` by `compendiumProtocolIdentifiers: SettingKey[Seq[ProtocolAndVersion]]`

Where `ProtocolAndVersion` could be:

```scala
final case class ProtocolAndVersion(name: String, version: Option[String])
```

.. or something like that. Does it make sense?

File: client/src/main/scala/sbtcompendium/client/CompendiumClient.scala [hunk 3]
```diff
@@ -102,7 +106,17 @@ object CompendiumClient {
...
+              .map(_.map(r => handleAvro(r.raw)).getOrElse(List.empty))
```
Comment (juanpedromoreno) [inline client/src/main/scala/sbtcompendium/client/CompendiumClient.scala]:
```suggestion
              .map(_.map(r => handleAvro(r.raw)).getOrElse(Nil))
```

File: client/src/main/scala/sbtcompendium/client/CompendiumClient.scala [hunk 4]
```diff
@@ -102,7 +106,17 @@ object CompendiumClient {
...
+          case _ =>
+            Sync[F].raiseError(UnknownError(s"Unknown error with status code 501. Schema format not implemented yet"))
```
Comment (juanpedromoreno) [inline client/src/main/scala/sbtcompendium/client/CompendiumClient.scala]: With context-applied, this could be just:

```suggestion
            F.raiseError(UnknownError(s"Unknown error with status code 501. Schema format not implemented yet"))
```

File: client/src/main/scala/sbtcompendium/client/CompendiumClient.scala [hunk 5]
```diff
@@ -102,7 +106,17 @@ object CompendiumClient {
...
+        }
+
+      private def handleAvro(raw: String): List[String] =
+        Generator(Standard).stringToStrings(raw)
```
Comment (juanpedromoreno) [inline client/src/main/scala/sbtcompendium/client/CompendiumClient.scala]: This can be considered an _effect_ that we might want to encapsulate in the F context.

Let me suggest you the following:

```scala
override def generateClient(target: IdlName, identifier: String): F[List[String]] =
        target match {
          case IdlName.Avro =>
            for {
              protocol <- retrieveProtocol(identifier, None)
              code     <- handleAvro(protocol)
            } yield code

          case _ =>
            F.raiseError(
              UnknownError(
                s"Unknown error with status code 501. " +
                  s"Schema format not implemented yet"
              )
            )
        }

      private def handleAvro(protocol: Option[Protocol]): F[List[String]] =
        protocol.map(r => F.delay(Generator(Standard).stringToStrings(r.raw))).getOrElse(F.pure(Nil))
```

File: plugin/src/main/scala/sbtcompendium/CompendiumPlugin.scala [hunk 6]
```diff
@@ -22,29 +22,34 @@ import hammock.asynchttpclient.AsyncHttpClientInterpreter
...
 
   override def trigger: PluginTrigger = allRequirements
 
   object autoImport {
 
-    lazy val compendiumGenClients: TaskKey[Seq[File]] = taskKey[Seq[File]]("Generate all the clients for each protocol")
-
-    lazy val compendiumServerHost: SettingKey[String] = settingKey[String]("Url of the compendium server")
-    lazy val compendiumServerPort: SettingKey[Int]    = settingKey[Int]("Port of the compendium server")
+    import sbtcompendium.client.{CompendiumClient, CompendiumClientConfig}
```
Comment (juanpedromoreno) [inline plugin/src/main/scala/sbtcompendium/CompendiumPlugin.scala]: Let's put all the imports at the top of the file

File: plugin/src/main/scala/sbtcompendium/CompendiumPlugin.scala [hunk 7]
```diff
@@ -22,29 +22,34 @@ import hammock.asynchttpclient.AsyncHttpClientInterpreter
...
+    lazy val compendiumGenClients: TaskKey[Seq[File]]   = taskKey[Seq[File]]("Generate all the clients for each protocol")
+    lazy val compendiumServerHost: SettingKey[String]   = settingKey[String]("Url of the compendium server")
+    lazy val compendiumServerPort: SettingKey[Int]      = settingKey[Int]("Port of the compendium server")
+    lazy val compendiumFormatSchema: SettingKey[String] = settingKey[String]("Schema type to download")
```
Comment (juanpedromoreno) [inline plugin/src/main/scala/sbtcompendium/CompendiumPlugin.scala]: Let's follow the same pattern as in https://github.com/higherkindness/mu-scala/blob/master/modules/srcgen/plugin/src/main/scala/higherkindness/mu/rpc/srcgen/SrcGenPlugin.scala#L36

File: plugin/src/main/scala/sbtcompendium/CompendiumPlugin.scala [hunk 8]
```diff
@@ -22,29 +22,34 @@ import hammock.asynchttpclient.AsyncHttpClientInterpreter
...
-      implicit val interpreter = AsyncHttpClientInterpreter.instance[cats.effect.IO]
-      implicit val clientConfig: CompendiumClientConfig = CompendiumClientConfig(
-        HttpConfig(host, port)
-      )
+      implicit val interpreter                          = AsyncHttpClientInterpreter.instance[cats.effect.IO]
+      implicit val clientConfig: CompendiumClientConfig = CompendiumClientConfig(HttpConfig(host, port))
       CompendiumClient[cats.effect.IO]()
     }
+
+    def fromStringToIdlName(sch: String) = sch match {
```
Comment (juanpedromoreno) [inline plugin/src/main/scala/sbtcompendium/CompendiumPlugin.scala]: You wouldn't need this function if we directly receive the type through the `compendiumFormatSchema` sbt setting

File: plugin/src/main/scala/sbtcompendium/CompendiumPlugin.scala [hunk 9]
```diff
@@ -53,19 +58,23 @@ object CompendiumPlugin extends AutoPlugin {
...
-        log.info(s"Attempting to generate client for [${protocolId}]")
-        generateProtocol
+      val generateProtocols = compendiumProtocolIdentifiers.value.toList.map {
+        protocolId =>
+          def targetFile(id: String) =
+            (sbt.Keys.sourceManaged in Compile).value / "compendium" / s"$protocolId$id.scala"
+          val generateProtocol = CompendiumUtils.generateCodeFor(
+            protocolId,
+            targetFile,
+            client(compendiumServerHost.value, compendiumServerPort.value).generateClient,
```
Comment (juanpedromoreno) [inline plugin/src/main/scala/sbtcompendium/CompendiumPlugin.scala]: Maybe we can extract the client creation outside of the `toList.map`, so we'd only need to create one client per compilation cycle (that's my bad :D )

File: project/ProjectPlugin.scala [hunk 10]
```diff
@@ -34,11 +34,13 @@ object ProjectPlugin extends AutoPlugin {
...
         %%("cats-core", V.cats),
-        "io.higherkindness"               %% "compendium-common" % V.compendium,
-        "com.pepegar"                     %% "hammock-core" % V.hammock,
-        "com.pepegar"                     %% "hammock-circe" % V.hammock,
-        "com.pepegar"                     %% "hammock-asynchttpclient" % V.hammock,
+        "io.higherkindness" %% "compendium-common"       % V.compendium,
+        "com.pepegar"       %% "hammock-core"            % V.hammock,
+        "com.pepegar"       %% "hammock-circe"           % V.hammock,
+        "com.pepegar"       %% "hammock-asynchttpclient" % V.hammock,
+        //"com.pepegar"                     %% "hammock-apache-http" % V.hammock,
```
Comment (juanpedromoreno) [inline project/ProjectPlugin.scala]: Let's remove this unused dependency

File: project/ProjectPlugin.scala [hunk 11]
```diff
@@ -34,11 +34,13 @@ object ProjectPlugin extends AutoPlugin {
...
         "com.beachape"                    %% "enumeratum" % V.enumeratum,
+        "com.julianpeeters"               %% "avrohugger-core" % "1.0.0-RC22",
```
Comment (juanpedromoreno) [inline project/ProjectPlugin.scala]: We should put the avrohugger version under `V` object, with the rest of the lib versions

File: client/src/main/scala/sbtcompendium/client/CompendiumClient.scala (hunk 4 above)
Comment (fedefernandez) [reply client/src/main/scala/sbtcompendium/client/CompendiumClient.scala]: Another option (since we have the `cats.implicits._` import) is just
```suggestion
            UnknownError(s"Unknown error with status code 501. Schema format not implemented yet").raiseError[F, List[String]]
```

File: plugin/src/sbt-test/sbtcompendium/basic/build.sbt [hunk 12]
```diff
@@ -2,6 +2,7 @@ import cats.effect.IO
 import cats.implicits._
 import higherkindness.compendium.models.IdlName
 import sbtcompendium.CompendiumUtils
+import cats.syntax._
```
Comment (fedefernandez) [inline plugin/src/sbt-test/sbtcompendium/basic/build.sbt]: Having the `implicits` this shouldn't be necessary

File: client/src/main/scala/sbtcompendium/client/CompendiumClient.scala (hunk 2 above)
Comment (ememmr) [reply client/src/main/scala/sbtcompendium/client/CompendiumClient.scala]: Yes, I thought the same but I preferred a simpler first version. However, it's ok for me to add it in this PR

File: client/src/main/scala/sbtcompendium/client/CompendiumClient.scala (hunk 4 above)
Comment (ememmr) [reply client/src/main/scala/sbtcompendium/client/CompendiumClient.scala]: the 
```UnknownError(s"Unknown error with status code 501. Schema format not implemented yet").raiseError[F, List[String]]```  
option looks cool 😎

File: client/src/main/scala/sbtcompendium/client/CompendiumClient.scala (hunk 5 above)
Comment (ememmr) [reply client/src/main/scala/sbtcompendium/client/CompendiumClient.scala]: agree, didn't realise

File: plugin/src/main/scala/sbtcompendium/CompendiumPlugin.scala (hunk 9 above)
Comment (ememmr) [reply plugin/src/main/scala/sbtcompendium/CompendiumPlugin.scala]: sure

File: plugin/src/sbt-test/sbtcompendium/basic/build.sbt [hunk 13]
```diff
@@ -7,34 +7,34 @@ scalaVersion := "2.12.10"
...
   )
 
-compendiumGenClients := {
+compendiumSrcGenClients := {
 
-  val generateProtocols = compendiumProtocolIdentifiers.value.toList.map { protocolId =>
-    val targetFile       = (sbt.Keys.sourceManaged in Compile).value / "compendium" / s"$protocolId.scala"
-    val generateProtocol = CompendiumUtils.generateCodeFor(protocolId, targetFile, generateClient)
+  val generateProtocols = compendiumSrcGenProtocolIdentifiers.value.toList.map { protocolId =>
+    def targetFile(id: String) = (sbt.Keys.sourceManaged in Compile).value / "compendium" / s"$protocolId.scala"
```
Comment (fedefernandez) [inline plugin/src/sbt-test/sbtcompendium/basic/build.sbt]: The `id` is ignored, isn't it? Can we revert it back to a `val`?

File: plugin/src/main/scala/sbtcompendium/CompendiumUtils.scala [hunk 14]
```diff
@@ -26,14 +26,20 @@ object CompendiumUtils {
...
+      path: String => File,
+      f: (IdlName, String) => IO[List[String]],
+      format: IdlName
+  ): Either[(String, Throwable), List[File]] =
+    f(format, identifier)
+      .map(
+        _.zipWithIndex
+          .map {
+            case (str, id) =>
+              val p = path("_class" + id.toString)
```
Comment (fedefernandez) [inline plugin/src/main/scala/sbtcompendium/CompendiumUtils.scala:38]: Why `_class`? do we need it?

**Review by fedefernandez — APPROVED** [review]

File: plugin/src/sbt-test/sbtcompendium/basic/build.sbt (hunk 13 above)
Comment (ememmr) [reply plugin/src/sbt-test/sbtcompendium/basic/build.sbt]: In the example is not needed but expected by the `generateCodeFor` function, but  in nested avro case, it generates one file per class so we need to add an id to not overwrite it

File: plugin/src/main/scala/sbtcompendium/CompendiumUtils.scala (hunk 14 above)
Comment (ememmr) [reply plugin/src/main/scala/sbtcompendium/CompendiumUtils.scala:38]: I added it to make it clearer. Since  it generates one file per case class

File: client/src/main/scala/sbtcompendium/client/CompendiumClient.scala [hunk 15]
```diff
@@ -102,7 +105,22 @@ object CompendiumClient {
...
+            } yield code
+          case _ =>
+            UnknownError(s"Unknown error with status code 501. Schema format not implemented yet")
+              .raiseError[F, List[String]]
+        }
+
+      private def handleAvro(op: Option[Protocol]): F[List[String]] =
+        op.map(p => F.delay(Generator(Standard).stringToStrings(p.raw))).getOrElse(F.pure(Nil))
+
+      private def safeInt(s: Option[String]): Option[Int] = s.flatMap(str => Try(str.toInt).toOption)
```
Comment (juanpedromoreno) [inline client/src/main/scala/sbtcompendium/client/CompendiumClient.scala:123]: The version could be "1.2.3", right? Or is compendium only supporting integers?

Comment (ememmr) [reply client/src/main/scala/sbtcompendium/client/CompendiumClient.scala:123]: Only integers

**Review by juanpedromoreno — APPROVED** [review]
